examples/grammar_constraint.py

```diff
@@ -0,0 +1,208 @@
+"""SMC Steering with Grammar Constraints
+
+Author: Gabriel Grand (grandg@mit.edu)
+
+This example illustrates grammar-constrained inference with SMC Steering.
+`GrammarConstrainedSMC` takes as input a grammar in Lark format.
+We use the Synchromesh (Poesia et al., 2022) to align the grammar with the
+language model vocabulary.
+
+Requires synchromesh (github.com/kanishkg/synchromesh)
+"""
+import asyncio
+import os
+from typing import List
+
+from hfppl.distributions import LMContext
+from hfppl.llms import CachedCausalLM
+from hfppl.modeling import Model
+from hfppl.inference import smc_standard
+
+from synchromesh.completion_engine import LarkCompletionEngine
+from synchromesh.synchromesh import StreamingCSD
+
+
+class GrammarConstrainedSMC(Model):
+    def __init__(
+        self,
+        lm: CachedCausalLM,
+        grammar: str,
+        start_rule: str,
+        prompt: str = None,
+        allow_ws: bool = False,
+        max_tokens: int = 32,
+        verbose: bool = False,
+    ):
+        super().__init__()
+        self.lm = lm
+        self.grammar = grammar
+        self.context = LMContext(lm, prompt)
+        self.vocab = self.lm.vocab
+
+        self.comp_engine = LarkCompletionEngine(
+            grammar, start_token=start_rule, allow_ws=allow_ws
+        )
+        self.csd = StreamingCSD(
+            completion_engine=self.comp_engine,
+            lm_vocabulary=self.vocab,
+            enforce_token_maximality=False,
+        )
+
+        self.max_tokens = max_tokens
+        self.n_tokens = 0
+
+        self.verbose = verbose
+
+    async def step(self):
+        # Get valid tokens for next step
+        valid_token_ids = self.csd.get_valid_tokens()
+
+        # Sample a token from the valid tokens
+        await self.observe(self.context.mask_dist(set(valid_token_ids)), True)
+        token = await self.sample(self.context.next_token())
+
+        # Feed the token to StreamingCSD
+        self.csd.feed_prediction(token.token_id)
+        self.n_tokens += 1
+
+        if self.verbose:
+            print(str(self.context))
+
+        # Check if generation is a full completion in the grammar
+        if self.csd.is_complete():
+            self.finish()
+        # Check for end of sentence
+        elif token.token_id == self.lm.tokenizer.eos_token_id:
+            self.finish()
+        # Max tokens reached
+        elif self.n_tokens >= self.max_tokens:
+            self.condition(False)
+            self.finish()
```

Comment:
Thanks, @gabegrand, this looks great! A couple questions about these lines:

1. Under what circumstances does Synchromesh say that `EOS` is a valid next token? It would make sense if this is only the case when `self.csd.is_complete()` is true -- do you know if that's the case?
2. We probably don't want to finish generation as soon as `self.csd.is_complete()` (unless I'm misunderstanding what `is_complete` does). This is because some valid strings may be prefixes of other valid strings in the grammar. Maybe we should only terminate if either (a) there are no valid next tokens and `self.csd.is_complete()` is true, or (b) the model outputs EOS and `self.csd.is_complete()` is true.

Reply:
Hey @alex-lew, great points. On closer inspection:

1. `csd.get_valid_tokens()` and `csd.is_complete()` appear to have no special handling for `EOS`: the underlying logic is based purely on incremental parsing according to the grammar and is agnostic to the LLM vocab. This has some important implications that I'll unpack below.
2. This is a very good catch. Agreed that we don't want to finish generation as soon as `csd.is_complete()`.

I've updated the code to follow your suggested termination logic. Note that I moved the "no valid next tokens" case to occur right after `csd.get_valid_tokens()`. This provides coverage for other potential instances where there are no valid next tokens (which would otherwise result in an error) and minimizes the number of separate calls to `csd.get_valid_tokens()`.

Now, with respect to EOS: As is, `csd.get_valid_tokens()` will never consider EOS a valid token (except by coincidence). Thus, currently, the logic effectively forces the LLM to keep generating until it either produces a valid completion in the grammar, exceeds the token limits, or runs out of options for next tokens under the grammar. I actually think this is a reasonable behavior and keeps the termination logic simple! Let's call this **Strategy 1**.

Alternatively, we could add EOS into the set of valid tokens at each generation step. Then, if the LLM samples EOS, we can call `self.condition(csd.is_complete())`. This would allow the LLM to effectively end generations that it considers low-probability, but could result in situations where we're never able to get a valid completion. Let's call this **Strategy 2**.

Currently, I've updated the code to follow Strategy 1. I ran a few quick tests with the example grammar and it seems to work. Let me know what you think makes sense here.

Reply:
Thanks Gabe! But why not put EOS into the set of valid tokens only when csd.is_complete()? That is, why does Strategy 2 need to use `condition`?

Reply:
Right -- this makes a lot of sense. I've updated the EOS termination logic to allow the LLM to finish generating if and only if `csd.is_complete()`. 

Note that now, all particles returned by the method are explicitly EOS-terminated; e.g., `'school of BIO433</s>'`. At least for explanatory purposes, I think this is desirable, as it makes clear that `p(EOS)` is part of the particle weight.

Thanks again for helping to think through this logic. Please take a look and let me know if this looks reasonable!

Reply:
Yep, that sounds great! Left one more question on another part of the code, but overall, this looks good to me.

Something we should try soon (but not necessarily in this PR) is to generate from this using a realistic grammar for e.g. Python or SQL. That should help us better understand the performance of this approach, and also surface any unintuitive behavior that we get from this approach.

Reply:
Great! Thanks again for your review. Reply to your comment below. 

Agreed, it will be good to test this out with a Python or SQL grammar and we can adjust accordingly. In the meantime, let me know if you think this example is ready to merge?